### PR TITLE
Scan/Kymo: add shape property to `Scan` and `Kymo`.

### PR DESCRIPTION
**Pull request by JoepVanlier**
**Why this PR?**
This PR is part of making `Scan`, `Kymo` and `CorrelatedStack` have similar properties. It's convenient to have a `shape` property.

Note that in the process of developing this feature I noticed two bugs that resulted in an error when downsampling or cropping by distance an image where some of the color channels were not defined. I fixed these and left the fixes as separate commits. If approved, I would keep those separate from the shape feature, but merge the two fixes together.

### CODE REVIEW

File: lumicks/pylake/scan.py [hunk 1]
```diff
@@ -352,8 +352,16 @@ def lines_per_frame(self):
 
     @property
     def _shape(self):
+        # Note that this is not the shape of the image property, but rather the shape in terms of
+        # [slow axis, fast axis].
```
Comment (tobiasjj) [inline lumicks/pylake/scan.py:356]: I appreciate the comment, but I still had a hard time to follow why `_shape()` is not the actual shape of the image. If I understand it corretly, `_shape()` returns the original shape of the `raw_image`, i.e. the uncropped image. Would it therefore make sense to rename the property `_shape` to `_raw_image_shape`, `_original_image_shape` or similar? I checked, it would require to change the name in three other places (`confocal._default_image_factory()`, `confocal._default_timestamp_factory()`, and `Kymo._shape`).

File: lumicks/pylake/kymo.py [hunk 2]
```diff
@@ -212,6 +212,20 @@ def _to_spatial(self, data):
     def _shape(self):
         return (self.pixels_per_line,)
```
Comment (tobiasjj) [inline lumicks/pylake/kymo.py:213]: Probably rename or add comment explaining, why `Kymo._shape` does not reflect the actual shape of the Kymo (see explanation in comment for property in `Scan._shape`).

**Review by tobiasjj — CHANGES_REQUESTED** [review]
I think the addition of the shape property will help users to work with `Kymo`s and `Scan`s. I have a few things:

1) Explicitly clarify that `_shape` refers to the original/raw image shape
2) Remove superfluous assignment in`Kymo.get_image()` 
3) Add a test to check that `get_image()` does not throw an error when trying to crop empty channels
4) Match shape of empty channels in `EmptyKymo` and `Kymo`

Sorry for so many remarks 😅!

File: lumicks/pylake/kymo.py [hunk 3]
```diff
@@ -603,11 +623,13 @@ def plot_rgb(self):
...
 
-    def _image(self):
-        return np.empty((self.pixels_per_line, 0))
+    def _image(self, channel):
+        shape = (self.pixels_per_line, 0, 3) if channel == "rgb" else (self.pixels_per_line, 0)
+        return np.empty(shape)
 
     def get_image(self, channel="rgb"):
-        return self._image()
+        im = self._image(channel)
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:631]: Huh good catch, must have been a rebasing mishap. Sorry 'bout that. Not sure how I managed to read over it 😅

File: lumicks/pylake/scan.py (hunk 1 above)
Comment (JoepVanlier) [reply lumicks/pylake/scan.py:356]: Nah, it's in a different axis order (slow to fast; reconstruction order). It's purely used in the reconstruction of images from raw photon counts and is an implementational detail in how they are reshaped.

They do respond to cropping actually, since they use properties that change, but it doesn't matter because a cropped/downsampled image is never reconstructed anew.

I hate the name too, so you're not alone. I think `_reconstruction_shape` would maybe be a more sensible name to convey that this is really a property that exists only for reconstruction purposes.

Maybe the docstring should read something like
```python
Shape used when reconstructing the image from raw photon counts (ordered by axis scan speed slow to fast).
``` 

What do you think?

File: lumicks/pylake/kymo.py [hunk 4]
```diff
@@ -212,6 +212,20 @@ def _to_spatial(self, data):
     def _shape(self):
         return (self.pixels_per_line,)
 
+    @property
+    def shape(self):
+        """Shape of the reconstructed `Kymo` image"""
+        # For a kymo the only way to find the shape is to perform the reconstruction. While one
+        # could traverse the info-wave for this purpose, we may as well just reconstruct the image
+        # since that still has the potential to be useful in the future (since it is cached).
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:220]: Yeah, I figured a reminder would be good. In the future, if we ever get the line count up front in the metadata, we can include a fast path to just pull it from metadata, because it's wasteful.

File: lumicks/pylake/kymo.py (hunk 2 above)
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:213]: Indeed. I'm all for changing the name.

File: lumicks/pylake/kymo.py [hunk 5]
```diff
@@ -443,7 +457,8 @@ def crop_by_distance(self, lower, upper):
         result._position_offset = self._position_offset + lower_pixels * self.pixelsize[0]
 
         def image_factory(_, channel):
-            return self._image(channel)[lower_pixels:upper_pixels, :]
+            data = self._image(channel)
+            return data[lower_pixels:upper_pixels, :] if data.size else data
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:461]: Yeah, I've added a test for this.

Thing is though, I think the entire missing channel handling can actually be overhauled to be much simpler (see last commit).

File: lumicks/pylake/kymo.py [hunk 6]
```diff
@@ -483,6 +498,11 @@ def downsampled_by(
 
         def image_factory(_, channel):
             data = self._image(channel)
+
+            # Make sure that missing color channels don't throw
+            if not data.size:
+                return data
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py]: I've added the test, but removed the statement for what I think is a better fix :D

File: lumicks/pylake/kymo.py [hunk 7]
```diff
@@ -603,11 +623,13 @@ def plot_rgb(self):
     def _plot(self, image, **kwargs):
         raise RuntimeError("Cannot plot empty kymograph")
 
-    def _image(self):
-        return np.empty((self.pixels_per_line, 0))
+    def _image(self, channel):
+        shape = (self.pixels_per_line, 0, 3) if channel == "rgb" else (self.pixels_per_line, 0)
+        return np.empty(shape)
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:626]: This should be fixed now.

File: lumicks/pylake/kymo.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobiasjj) [reply lumicks/pylake/kymo.py]: Dang it, you were faster with the fixup than I with double checking this one :-)

**Review by tobiasjj — APPROVED** [review]
Great! The overhaul of this PR with the idea of using the `infowave.data.size` to return pixels with no counts solves all the issues I had before and further even simplifies and "consistencyfies" the code in other places.
Only small nitpicks, otherwise good to go from my side.

File: lumicks/pylake/kymo.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py]: :D

File: lumicks/pylake/detail/confocal.py [hunk 8]
```diff
@@ -436,8 +440,9 @@ def infowave(self):
         return self.file["Info wave"]["Info wave"][self.start : self.stop]
 
     @property
-    def _shape(self):
-        """The shape of the image ([optional: pixels on slow axis], pixels on fast axis)"""
+    def _reconstruction_shape(self):
+        """The shape of the image ([optional: pixels on slow axis], pixels on fast axis). This
+        property is purely used for reconstruction from photon streams."""
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/detail/confocal.py:445]: :+1:

File: lumicks/pylake/detail/confocal.py [hunk 9]
```diff
@@ -531,15 +536,6 @@ def get_image(self, channel="rgb") -> np.ndarray:
...
-            try:
-                # Get first non-zero shape
-                image_shape = next(image.shape for image in images if image.size)
-                return np.stack(
-                    [image if image.size else np.zeros(image_shape) for image in images],
-                    axis=-1,
-                )
-            except StopIteration:
-                raise ValueError("No image data available")
+            return np.stack([self.get_image(color) for color in ("red", "green", "blue")], axis=-1)
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/detail/confocal.py:539]: 👍

File: lumicks/pylake/kymo.py [hunk 10]
```diff
@@ -443,7 +459,8 @@ def crop_by_distance(self, lower, upper):
         result._position_offset = self._position_offset + lower_pixels * self.pixelsize[0]
 
         def image_factory(_, channel):
-            return self._image(channel)[lower_pixels:upper_pixels, :]
+            data = self._image(channel)
+            return data[lower_pixels:upper_pixels, :]
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymo.py:463]: Nope, contracting 👍
